### PR TITLE
MCO-1276, MCO-1339: Integrate Metric Collection with MCO Daemon and e2e testing automation

### PR DESCRIPTION
**Pull request by dkhater-redhat**
**- What I did**
1. Created an e2e test to verify the installation of unsupported RPM packages on nodes using rpm-ostree.
2. Implemented metric collection in the mcd to report unsupported packages by vendor via the mcd_local_unsupported_packages_by_vendor metric.
3. Ensured that the e2e test verifies the installation of the unsupported package, checks that the correct metrics are being reported, and includes cleanup logic to uninstall the package after the test completes.

**- How to verify it**
Run the provided e2e test TestInstallRPMAndCheckMCDMetrics to verify that:
1. The RPM package (aeskeyfind) is successfully installed on the node.
2. The MCD reports the correct metric reflecting the installed package.
**- Description for the changelog**
Added e2e test for verifying unsupported RPM package installation and metric reporting in the mcd.

### CODE REVIEW

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:2836]: Might be good to have this and our supported extensions share a constant somewhere, so we don't have to update both in case we add more packages

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:2865]: nit: we can probably remove some of these logs

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:2862]: I think if we don't find the vendor, we can probably still add it to the list of packages (i.e. just set vendor to empty but don't `continue` here, just so we can add that to the metrics regardless. WDYT?

File: test/e2e/mcd_test.go
Comment (yuqi-zhang) [inline test/e2e/mcd_test.go:1133]: We discussed this and I think we should be ok with this, since we'll likely remove this alert after we move to layering GA.

Just curious, does the multi arch build work?

**Review by yuqi-zhang — COMMENTED** [review]
Generally makes sense to me! some questions/minor suggestions inline

File: test/e2e/mcd_test.go
Comment (dkhater-redhat) [reply test/e2e/mcd_test.go:1133]: yes so there are packages that have "noarch" that should work. I didn't try them because I found the x86 version of aeskeyfind which ended up working. I could change it to a noarch one if needed but x86 should be fine for CI purposes.

File: pkg/daemon/daemon.go
Comment (dkhater-redhat) [reply pkg/daemon/daemon.go:2862]: yeah i agree- good catch

File: pkg/daemon/daemon.go
Comment (dkhater-redhat) [reply pkg/daemon/daemon.go:2836]: great- just saw a getSupportedExtensions() in the daemon so will use that.

File: test/e2e/mcd_test.go
Comment (cheesesashimi) [inline test/e2e/mcd_test.go:1251]: thought (non-blocking): I vaguely recall a similar function to this elsewhere within our codebase. In the future, it might be useful to make a more general-purpose version of this code.

File: test/helpers/utils.go
Comment (cheesesashimi) [inline test/helpers/utils.go:383]: thought (non-blocking): I vaguely recall there being a helper function very similar to this, but I'm not 100% sure.

File: test/e2e/mcd_test.go
Comment (cheesesashimi) [inline test/e2e/mcd_test.go:1133]: suggestion: Put this text string into a separate file and use Go embedding to read it in.

File: test/helpers/utils.go
Comment (dkhater-redhat) [reply test/helpers/utils.go:383]: I thought so, too, but I couldn't find one looking through the helpers. I did end up adding the `WaitForNodeReady` helper but I may have overlooked it. I just parsed through the helpers again and didn't see anything that would be a great fit (unless I overlooked it again)

File: test/e2e/mcd_test.go
Comment (cheesesashimi) [inline test/e2e/mcd_test.go:1162]: issue: This MachineConfig gets rolled out to all of the worker nodes which can slow things down

suggestion: Create a MachineConfigPool just for this test, move the desired node into it, and apply the MachineConfig to just that node. You can use this helper to do all of that for you: hhttps://github.com/openshift/machine-config-operator/blob/master/test/helpers/utils.go#L135 and as a bonus, it has a cleanup function that will revert everything in the correct order.

File: test/e2e/mcd_test.go
Comment (cheesesashimi) [inline test/e2e/mcd_test.go:1173]: thought: You don't need to use a MachineConfig to get the RPM onto the node. Instead, you can do something like this: `$ oc cp "$PWD/file.rpm" machine-config-daemon-<id>:/rootfs/tmp/file.rpm -c machine-config-daemon -n openshift-machine-config-operator`

You'll need to identify the correct MCD for the node, but there are helpers for that. Look at how ExecCmdOnNode works to get an idea for that.

Comment (dkhater-redhat) [reply test/e2e/mcd_test.go:1173]: im going with the MCP route in your previous comment- do you think its MC application related?  just confused where the context deadlines are coming from

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:2823]: After some slack discussion I think we don't need to check for Modules. I think it's fine to remove this and track everything else.

**Review by yuqi-zhang — COMMENTED** [review]
Generally looks good! One suggestion inline.

Also, the commits look a bit weird. One commit still adds the test rpms and the second removes them. Could you restructure the commits two:
1. tweaking the metric (daemon.go and metrics.go)
2. adding tests (mcd_test.go and utils.go)

To make it easier for backport, and so that we don't have the weird rpm history. Thanks!

**Comment by dkhater-redhat** [comment]
thank you @yuqi-zhang and @sergiordlr 

sergio, would you mind taking another look at this when you get a chance?

**Review by yuqi-zhang — APPROVED** [review]
/lgtm
/hold

Will let Sergio check again before merging. Please unhold if the functionality looks good, thanks!

**Comment by sergiordlr** [comment]
Hello! I see that the vendor has been removed. There is no way to know which kind of packages were modified

In https://issues.redhat.com/browse/MCO-1275 we can read:
```
Acceptance Criteria:
1. List of metrics that need to be collected, including installed packages that aren't CoreOS extensions.
2. Clear definition of how to identify packages that were installed via rpm-ostree install $pkg.
```

Are we OK with not getting any information about the modified rpms?

**Comment by dkhater-redhat** [comment]
hello @sergiordlr thanks for looking at this!
1. we have opted to collect metrics per node instead of per vendor. this is intended functionality.
2. we should probably fix the unsupported package functionality around supported ext. 
3. good find on the MCO ext not allowing further ext installs. we should probably look into that more. although I do wonder (tangentially)- would this work also be further enabling the customer to install ext that are unsupported? just a thought

**Comment by dkhater-redhat** [comment]
alright @sergiordlr this is fixed! please review once more that those packages are seen and we should be gtg :-D

**Comment by sergiordlr** [comment]
Hello! everything seems to be working fine

Could you confirm that we are not going to support this?

```
Not reported:
# requested-base-remote-replacements
sh-5.1# rpm-ostree override replace --experimental  --from repo=baseos vim-minimal

sh-5.1# rpm-ostree status
State: idle
Deployments:
  ostree-unverified-registry:registry.build09.ci.openshift.org/ci-ln-2396jjt/stable@sha256:0046122617324ad3105d1c76655841864aea009ad87f3ff9adbb13a4a4debf7e
                   Digest: sha256:0046122617324ad3105d1c76655841864aea009ad87f3ff9adbb13a4a4debf7e
                  Version: 418.94.202410311703-0 (2024-10-31T17:08:09Z)
                     Diff: 1 upgraded
          RemoteOverrides: repo=baseos
                           `- vim-minimal 2:8.2.2637-20.el9_1 -> 2:8.2.2637-21.el9


```

**Comment by dkhater-redhat** [comment]
@sergiordlr we will not be supporting that functionality. This could be a future consideration though. thanks for looking out!

**Comment by sergiordlr** [comment]
Everything is working as expected. We can add the qe-approved label. 

/label qe-approved

Thank you very much @dkhater-redhat !!

**Pull request merged**
